iceoryx_dds/test/moduletests/test_fastdds.cpp

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_dds/dds/fastdds_data_reader.hpp"
+#include "iceoryx_dds/dds/fastdds_data_writer.hpp"
+#include "test.hpp"
+
+#include <fastrtps/xmlparser/XMLProfileManager.h>
+
+#include <chrono>
+
+using namespace ::testing;
+using ::testing::_;
+
+namespace iox
+{
+namespace dds
+{
+// ======================================== Helpers ======================================== //
+using TestDataReader = FastDDSDataReader;
+using TestDataWriter = FastDDSDataWriter;
+
+// ======================================== Fixture ======================================== //
+class FastDDSTest : public Test
+{
+  public:
+    void SetUp(){};
+    void TearDown(){};
+};
+
+// ======================================== Tests ======================================== //
+TEST_F(FastDDSTest, DoesNotAttemptToReadWhenDisconnected)
+{
+    // ===== Setup
+    uint64_t bufferSize = 1024;
+    uint8_t buffer[bufferSize];
+
+    // ===== Test
+    TestDataReader reader{"", "", ""};
+
+    auto takeResult = reader.take(buffer, bufferSize, iox::cxx::nullopt);
+    EXPECT_EQ(true, takeResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::NOT_CONNECTED, takeResult.get_error());
+
+    auto takeNextResult = reader.takeNext(buffer, bufferSize);
+    EXPECT_EQ(true, takeNextResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::NOT_CONNECTED, takeResult.get_error());
+}
+
+TEST_F(FastDDSTest, ReturnsErrorWhenAttemptingToReadIntoANullBuffer)
+{
+    // ===== Setup
+    uint64_t bufferSize = 0;
+    uint8_t* buffer = nullptr;
+
+    // ===== Test
+    TestDataReader reader{"", "", ""};
+    reader.connect();
+
+    auto takeResult = reader.take(buffer, bufferSize, iox::cxx::nullopt);
+    EXPECT_EQ(true, takeResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::INVALID_RECV_BUFFER, takeResult.get_error());
+
+    auto takeNextResult = reader.takeNext(buffer, bufferSize);
+    EXPECT_EQ(true, takeNextResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::INVALID_RECV_BUFFER, takeNextResult.get_error());
+}
+
+TEST_F(FastDDSTest, SendAndTakeNext)
```

Comment:
Could you maybe also add a test where the bufferSize is a prime number like `1021`. I think you may encounter issues since you may have a rounding error in `fastdds_data_reader.cpp` (`auto bufferCapacity = bufferSize / sampleSize`) ... See comment.